### PR TITLE
[python][torch] Add lazy contiguous window dataset

### PR DESCRIPTION
**Pull request by YannByron**
## Summary

Add a generic, snapshot-pinned `ContiguousWindowDataset` for exposing fixed-size Paimon row windows to PyTorch without loading projected rows or BLOB payloads into memory up front.

This is the contiguous-window portion split from #9466 following review feedback. The ACT benchmark workflow will be submitted separately.

## Changes

- Add row-ID-backed window indexing with deterministic group/order semantics, configurable stride, and drop/pad/error tail policies.
- Read projected values lazily from the pinned snapshot, including scalar and `MAP<..., BLOB>` payloads and anchor-only columns.
- Coalesce overlapping plural reads while keeping logical samples independent for transforms and adapters.
- Expose the Dataset through `ScanQuery.to_contiguous_window_dataset` and document its PyTorch usage and contracts.
- Cover snapshot pinning, BLOB handling, batching, mutable values, padding, pickling, and multi-worker DataLoader behavior.

## Testing

- [x] `PYTHONPATH=. /opt/homebrew/bin/python3.13 -m pytest pypaimon/tests/contiguous_window_dataset_test.py -q` (16 passed)
- [x] `PYTHONPATH=. /opt/homebrew/bin/python3.13 -m pytest pypaimon/tests/multimodal_table_test.py -k 'scan_read_blobs or scan_read_and_stream_map_blobs' -q` (6 passed, 74 deselected)
- [x] `/opt/homebrew/bin/python3.13 -m flake8 --config=dev/cfg.ini pypaimon/multimodal/query.py pypaimon/multimodal/window_dataset.py pypaimon/tests/contiguous_window_dataset_test.py`
- [x] `python3 dev/check_license_header.py`
- [x] `/opt/homebrew/bin/python3.13 -m compileall -q pypaimon/multimodal`
- [x] `git diff --check origin/master..HEAD`

## Notes

No ACT benchmark implementation, packaging, tests, or benchmark documentation is included in this PR.

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Requirement fit is supported, but the inline findings below should be addressed before merging.

File: paimon-python/pypaimon/multimodal/query.py [hunk 1]
```diff
@@ -164,6 +164,70 @@ def to_torch(
...
+            column_transforms: Per-column callables applied to value lists.
+            pad_values: Optional replacement values used by ``tail='pad'``.
+            adapter: Callable that converts the complete sample mapping.
+            blob_parallelism: Maximum concurrent BLOB body reads per fetch.
+
+        Returns:
+            A snapshot-pinned ``ContiguousWindowDataset``. See that class for
+            padding, mask, transform, and adapter result semantics.
+        """
+        if self._result_factory is not None:
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/query.py]: [P1] Reject batch-vector queries instead of silently scanning the base table

This scan-only guard relies on `_result_factory`, but `BatchVectorQuery` deliberately leaves `_result_factory=None` because it implements search through its overridden `to_arrow()`. Consequently, `search_vectors(...).to_contiguous_window_dataset(...)` passes this guard, `_read_window_index()` uses the ordinary scan builder, and the resulting Dataset contains full-table rows while ignoring the vector results and `_pre_filter`. Please reject this method in `_PreFilterQuery`, alongside the existing scan-only API overrides, and add a `search_vectors()` regression test. The public `from_query` path should enforce the same query-kind check.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+        CoreOptions.SCAN_MODE,
+        CoreOptions.INCREMENTAL_BETWEEN_TIMESTAMP,
+        CoreOptions.SCAN_FILE_CREATION_TIME_MILLIS,
+        CoreOptions.SCAN_CREATION_TIME_MILLIS,
+    ))
+    options = {
+        key: None for key in scan_keys
+        if table.options.options.contains_key(key)
+    }
+    options[CoreOptions.SCAN_SNAPSHOT_ID.key()] = str(snapshot_id)
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/window_dataset.py]: [P1] Preserve reads through tag-retained snapshot metadata

Replacing every resolved `scan.tag-name` with `scan.snapshot-id` loses the tag's retention namespace. After the main snapshot file expires, the tagged scan remains valid and can still build this index, but the first deferred item read fails with `ValueError: Snapshot id '1' doesn't exist.` Please keep subsequent reads on the tag-retained snapshot (with resolved-ID validation), or introduce a read path that can use the resolved Tag/Snapshot directly, and cover a retained tag whose main snapshot file has expired.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+            columns: Projected value columns, or all Dataset columns when
+                omitted.
+
+        Returns:
+            A list of row dictionaries aligned one-for-one with ``row_ids``.
+            The internal ``_ROW_ID`` field is removed, and BLOB descriptors are
+            resolved to their bodies.
+        """
+        columns = self.columns if columns is None else columns
+        query = ScanQuery(self._table)
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/window_dataset.py]: [P2] Reuse the pinned scan plan instead of replanning every fetch

Every `_read_rows()` call constructs a new query and executes `plan()` through either `to_arrow()` or `read_blobs()`. This means one full snapshot/manifest planning pass per direct item or modern DataLoader batch, and two when `anchor_columns` is configured; the work repeats on every epoch and in every worker. I instrumented a two-row table and each repeated item read added two manifest-list reads and one manifest-file read. Please retain authorized pinned-snapshot splits and route requested row-ID ranges to cached/`IndexedSplit` splits, as the existing lazy `TorchDataset` does.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+        return [by_row_id[row_id] for row_id in row_ids]
+
+
+def _read_window_index(query, group_key, order_key):
+    index_query = copy.copy(query)
+    index_query._projection = [group_key, order_key]
+    index_query._include_row_id = True
+    read_builder = index_query._configured_read_builder()
+    plan = read_builder.new_scan().plan()
+    index = read_builder.new_read().to_arrow(plan.splits())
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/window_dataset.py]: [P2] Do not treat authorization-masked row IDs as physical row IDs

This table is the authorization-processed plan output, so `_ROW_ID` may be masked. With the supported `_ROW_ID -> NULL` rule, ordinary projected scans still work, but Dataset construction reaches `int(None)` and misleadingly reports that the group key is unhashable. The existing lazy `TorchDataset` explicitly disables row-ID-backed access when a `QueryAuthSplit` masks `_ROW_ID`. Please inspect the plan for this condition and either use a safe authorized materialized fallback or reject it with an explicit unsupported/permission error; do not bypass the mask to obtain raw IDs.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+                group_values, order_values, row_ids):
+            if group_key is None:
+                raise ValueError("%s must not contain null values." % self.group_key)
+            if order_value is None:
+                raise ValueError("%s must not contain null values." % self.order_key)
+            if isinstance(order_value, bool) or not isinstance(order_value, Integral):
+                raise ValueError(
+                    "%s must contain integer values." % self.order_key)
+            try:
+                grouped[group_key].append((int(order_value), int(row_id)))
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/window_dataset.py]: [P2] Reject or canonicalize NaN group keys before dictionary grouping

Separate FLOAT/DOUBLE NaN values are non-null and hashable, but `nan != nan`, so using them directly as dictionary keys splits rows carrying the same NaN group value into separate groups. With two consecutive rows and `window_size=2`, I reproduced `tail='drop'` returning no windows and `tail='pad'` returning two separately padded groups, while Arrow `group_by` groups those NaNs together. Please reject non-reflexive group values with a clear error, or canonicalize NaNs before grouping, and add FLOAT/DOUBLE coverage.

File: docs/docs/pypaimon/multimodal-api.mdx [hunk 6]
```diff
@@ -866,6 +866,64 @@ Notes:
...
+### Contiguous windows for PyTorch
+
+Install the `torch` extra, then use `to_contiguous_window_dataset` to expose
+map-style windows without loading the selected rows or BLOB payloads into Python
+memory up front. The Dataset builds a compact index from the group column, order
+column, and Paimon row IDs. Each `__getitem__` call fetches only that window from
+the snapshot recorded in `dataset.snapshot_id`.
+
+```shell
+pip install pypaimon[torch]
```
Comment (JingsongLi) [inline docs/docs/pypaimon/multimodal-api.mdx]: [P2] Quote the Torch extra in this installation command

Default zsh interprets `pypaimon[torch]` as a glob and aborts with `no matches found` before pip runs. Please match the existing PyTorch documentation and write `pip install 'pypaimon[torch]'`.

**Comment by JingsongLi** [comment]
Regarding the schema design, we need to align our design and trade-offs with https://github.com/apache/paimon/pull/9529.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+from torch.utils.data import Dataset
+
+from pypaimon.common.options.core_options import CoreOptions
+from pypaimon.multimodal.query import ScanQuery
+from pypaimon.schema.data_types import is_blob_type, is_map_blob_type
+from pypaimon.snapshot.time_travel_util import SCAN_KEYS
+from pypaimon.table.special_fields import SpecialFields
+
+
+class ContiguousWindowDataset(Dataset):
```
Comment (XiaoHongbo-Hope) [inline paimon-python/pypaimon/multimodal/window_dataset.py:44]: Suggestion: WindowDataset, SequenceDataset or SequenceWindowDataset?

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+        query._projection = list(columns)
+        query._include_row_id = True
+
+        blob_columns = [
+            field.name for field in self._table.fields
+            if field.name in columns
+            and (is_blob_type(field.type) or is_map_blob_type(field.type))
+        ]
+        if blob_columns:
+            scalar, blobs = query.read_blobs(
```
Comment (XiaoHongbo-Hope) [inline paimon-python/pypaimon/multimodal/window_dataset.py]: This loses the `frame_index` of `.video` values.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+        Args:
+            index: Arrow table containing ``group_key``, ``order_key``, and
+                Paimon's ``_ROW_ID`` for the resolved snapshot.
+
+        Returns:
+            ``(groups, anchors)``. Each group stores its key, ordered positions,
+            and row IDs. Each anchor stores group index, start offset, and the
+            number of real rows available before optional padding.
+        """
+        group_values = index.column(self.group_key).to_pylist()
```
Comment (XiaoHongbo-Hope) [inline paimon-python/pypaimon/multimodal/window_dataset.py]: We can use Arrow or NumPy arrays instead to avoid memory issue?

File: docs/docs/pypaimon/multimodal-api.mdx [hunk 10]
```diff
@@ -866,6 +866,64 @@ Notes:
...
+    return torch.tensor(values, dtype=torch.float32)
+
+
+windows = (
+    frames.scan()
+    .where("split = 'train'")
+    .to_contiguous_window_dataset(
+        window_size=16,
+        columns=["state", "action"],
+        group_key="episode_id",
```
Comment (XiaoHongbo-Hope) [inline docs/docs/pypaimon/multimodal-api.mdx]: These defaults do not match the native LeRobot schema preserved by #9529, which uses episode_index and frame_index.

**Comment by JingsongLi** [comment]
I suggest defining RoboMIND's native Paimon storage contract early, following the approach in #9529:

- Preserve original `master`/`puppet` fields, dtypes and image/depth payloads in `frames`; keep trajectory metadata in `episodes`.
- Add `tasks`, `annotations` and `calibrations` when provided by the source.
- Publish components under the same `BIGINT version_id` tag, with a `versions` manifest and `READY` written last.
- Treat canonical state/action and normalization statistics as explicitly derived data, not as the native schema.

Start with AgileX without assuming other RoboMIND variants share its layout or sampling axis. This can be a separate schema/ingestion change; the window API should stay generic and be tested against that native contract.

File: paimon-python/pypaimon/multimodal/window_dataset.py (hunk 7 above)
Comment (YannByron) [reply paimon-python/pypaimon/multimodal/window_dataset.py:44]: I prefer to keep `ContiguousWindowDataset`: strict contiguity within each group is its defining guarantee. `WindowDataset` could imply arbitrary, non-contiguous windows, while `SequenceDataset` does not describe the map-style window sample returned by `__getitem__`.

File: paimon-python/pypaimon/multimodal/window_dataset.py (hunk 8 above)
Comment (YannByron) [reply paimon-python/pypaimon/multimodal/window_dataset.py]: Thanks for pointing out this additional case. The current RoboMIND HDF5 ingestion is image-based: it stores individual RGB/depth frames and has no video asset or video-frame-offset semantics. Therefore, this PR has not yet designed the window representation around `.video` descriptors.

We will treat video support as a follow-up extension and preserve `VideoFrameDescriptor.frame_index` and the other descriptor metadata when adding it. Until then, `.video` columns should be explicitly unsupported rather than silently losing metadata.

**Comment by YannByron** [comment]
> I suggest defining RoboMIND's native Paimon storage contract early, following the approach in #9529:
> 
> * Preserve original `master`/`puppet` fields, dtypes and image/depth payloads in `frames`; keep trajectory metadata in `episodes`.
> * Add `tasks`, `annotations` and `calibrations` when provided by the source.
> * Publish components under the same `BIGINT version_id` tag, with a `versions` manifest and `READY` written last.
> * Treat canonical state/action and normalization statistics as explicitly derived data, not as the native schema.
> 
> Start with AgileX without assuming other RoboMIND variants share its layout or sampling axis. This can be a separate schema/ingestion change; the window API should stay generic and be tested against that native contract.

[robomind_agilex](https://github.com/apache/paimon/blob/master/paimon-python/pypaimon/sample/robomind_agilex.py) already defines an explicit Paimon storage contract for the RoboMIND AgileX dataset through `episode_schema`, `frame_schema`, and `feature_stats_schema`. It covers dataset metadata, ordered frame data, image/depth payloads, and the normalization data required for training. The AgileX contract does not yet model `tasks`, `annotations`, or `calibrations`. 

The current `contiguous-window` dataset API addresses the requirements of the HDF5-based RoboMIND AgileX dataset while remaining extensible for future sources and modalities.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 11]
```diff
@@ -0,0 +1,707 @@
...
+        self._projection = (
+            [name for name in columns if name not in blob_column_set]
+            + [SpecialFields.ROW_ID.name]
+            + self._blob_columns
+        )
+        self._table = (
+            table.copy({CoreOptions.BLOB_AS_DESCRIPTOR.key(): "true"})
+            if self._blob_columns else table
+        )
+        self._splits = self._new_read_builder().new_scan().plan().splits()
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/window_dataset.py]: [P1] Revalidate the retained tag on every deferred plan

`_pin_table` verifies the tag only during dataset construction, but this lazy plan resolves the mutable tag name again and immediately discards `plan.snapshot_id`. `FileStoreTable.replace_tag` can therefore move the tag between those two points. I reproduced a dataset whose recorded `snapshot_id` stayed at snapshot 1 while the first lazy value read returned data updated only in snapshot 2. Pickled/DataLoader workers are also exposed because `__getstate__` clears the plans.

Please pass the expected snapshot ID into `_PinnedRowIdPlan`, keep the full `Plan`, and fail closed unless `plan.snapshot_id` still equals the indexed snapshot before caching its splits (or plan from an immutable resolved snapshot). A regression test should replace the tag after construction and before the first projection read, including the post-pickle path.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 12]
```diff
@@ -0,0 +1,707 @@
...
+    )).astype(np.int64, copy=False)
+
+
+def _read_window_index(query, group_key, order_key):
+    index_query = copy.copy(query)
+    index_query._projection = [group_key, order_key]
+    index_query._include_row_id = True
+    read_builder = index_query._configured_read_builder()
+    plan = read_builder.new_scan().plan()
+    splits = plan.splits()
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/window_dataset.py:547]: [P2] Reject authorization masks on the grouping and ordering keys

The index scan applies column masking, but the validation below rejects only a masked `_ROW_ID`. A non-injective mask on `group_key` can merge real groups before sorting and anchor construction; a mask on `order_key` can similarly change contiguity. I reproduced two groups `a:[0,1]` and `b:[2,3]` with a constant mask on `episode`, which produced a window crossing the group boundary (`[0,1,102]`) despite the API guarantee that windows never cross groups.

Please inspect `QueryAuthSplit.auth_result.column_masking` and reject masks on `group_key` or `order_key` unless the index can be built from unmasked keys without weakening authorization. Add a regression test using a non-null constant mask; null-only tests would be caught by the existing null validation and miss this case.

File: paimon-python/pypaimon/multimodal/window_dataset.py [hunk 13]
```diff
@@ -0,0 +1,707 @@
...
+            pruned for pruned in (
+                _prune_split_files(split, ranges) for split in self._splits)
+            if pruned is not None
+        ]
+        if not splits:
+            return []
+        read_builder = self._new_read_builder()
+        predicate = read_builder.new_predicate_builder().is_in(
+            SpecialFields.ROW_ID.name, requested)
+        arrow = read_builder.with_filter(predicate).new_read().to_arrow(splits)
```
Comment (JingsongLi) [inline paimon-python/pypaimon/multimodal/window_dataset.py:447]: [P2] Preserve exact row ranges when reading cached splits

The cached plan contains unfiltered `DataSplit`s, and `_prune_split_files` only drops whole files. This later `_ROW_ID IN (...)` predicate cannot recover native row-range pushdown for a data-evolution read: `DataEvolutionSplitRead._push_down_predicate()` deliberately returns `None`, and its `row_ranges` remains `None` unless the supplied split is an `IndexedSplit`. As a result, every overlapping file is decoded before `FilterRecordBatchReader` (and the Arrow filter below) removes unrelated rows.

That is on the documented shuffled training hot path: one batch can request many scattered windows and repeatedly decode large files. Please build a split-range index once, intersect requested ranges with matching splits, and pass `IndexedSplit`s while preserving `QueryAuthSplit`; `TorchDataset._SplitRangeIndex` / `_select_splits` is an existing implementation to reuse. Add a multi-file test that observes format-reader row ranges or decoded-row counts, since the six-row logical-call tests do not expose full-file decoding.

**Comment by JingsongLi** [comment]
+1
